add plugin to produce `FEDRawDataCollection` excluding a list of FEDs

missirol opened the pull request:
#### PR description:

This PR aims to introduce a producer that, given a `FEDRawDataCollection`, produces a new one _excluding_ a certain list of FEDs. This new plugin is, in a way, the counterpart of `EvFFEDSelector`, which instead requires the list of FEDs to keep.

One use case for this plugin is HIon's "RawPrime", which includes a `FEDRawDataCollection` without SiStrip FEDs (right now, an hard-coded list of "all FEDs but the SiStrip ones" is used in a `EvFFEDSelector`; it would arguably be better to have a `EvFFEDExcluder` with the list of SiStrip FEDs).

This was discussed in https://github.com/cms-sw/cmssw/pull/39863#discussion_r1006485795 and https://github.com/cms-sw/cmssw/pull/39863#discussion_r1013955672.

Names (e.g. "excluder") are tentative. I chose different names for the plugin's parameters with respect to `EvFFEDSelector` (`InputTag -> src`, `fedList -> fedsToExclude`) thinking that this reduces the chance of using `EvFFEDExcluder` instead of `EvFFEDSelector` (or viceversa) by mistake.

A cfg file is added to `test/` for testing purposes (if not useful, it can be removed).

No changes expected, as this PR is basically insensitive to PR tests.

_Note_ : no backports are planned. The intention is simply to add this new plugin for development purposes (could be useful in 2023).

FYI: @mmusich

#### PR validation:

Checks based on `testEvFFEDSelectors.py`.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

N/A

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39993/32923

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @missirol (Marino Missiroli) for master.

It involves the following packages:

- EventFilter/Utilities (**daq**)


@cmsbuild, @emeschi, @smorovic can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

smorovic commented:
please test

smorovic commented:
It looks fine and is more practical than providing a full list of FEDs for filtering (something can be accidentally skipped, too).

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c8184/28801/summary.html
**COMMIT**: b1a53241d08928203e6a73a15a9ba73932bc345b
**CMSSW**: CMSSW_12_6_X_2022-11-04-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39993/28801/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416402
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416374
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-04-2300+0c8184/53914/triggerResults): no differences found

smorovic commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

missirol commented:
hold

Let me push a small improvement: we can compute the list of selected FEDs already in the constructor, rather than `find`ing it in the event loop.

cmsbuild commented:
Pull request has been put on hold by  @missirol
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

missirol committed:
added plugin EvFFEDExcluder

The producer EvFFEDExcluder takes in input
a FEDRawDataCollection (1), and a list of FED IDs (2).

It puts in the Event a FEDRawDataCollection which is
a copy of (1) excluding the data buffers of the FED IDs listed in (2).

EvFFEDExcluder is, in a way, the counterpart of EvFFEDSelector,
which instead requires the list of FEDs to keep.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39993/32924

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #39993 was updated. @cmsbuild, @emeschi, @smorovic can you please check and sign again.

missirol commented:
unhold

missirol commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c8184/28804/summary.html
**COMMIT**: 00a6f0a0d337177dca4aaef8612735b85907cf09
**CMSSW**: CMSSW_12_6_X_2022-11-05-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39993/28804/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416402
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416374
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-05-1100+0c8184/53917/triggerResults): no differences found

smorovic commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented on EventFilter/Utilities/plugins/EvFFEDExcluder.cc:
```diff
@@ -0,0 +1,48 @@
+#include <algorithm>
+
+#include "EvFFEDExcluder.h"
+#include <FWCore/Framework/interface/Event.h>
+#include <FWCore/ParameterSet/interface/ConfigurationDescriptions.h>
+#include <FWCore/ParameterSet/interface/ParameterSet.h>
+#include <FWCore/ParameterSet/interface/ParameterSetDescription.h>
+#include <FWCore/Utilities/interface/InputTag.h>
+#include <DataFormats/FEDRawData/interface/FEDNumbering.h>
+#include <DataFormats/FEDRawData/interface/FEDRawDataCollection.h>
```
```suggestion
#include "FWCore/Framework/interface/Event.h"
#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
#include "FWCore/ParameterSet/interface/ParameterSet.h"
#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
#include "FWCore/Utilities/interface/InputTag.h"
#include "DataFormats/FEDRawData/interface/FEDNumbering.h"
#include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
```

missirol replied:
Fixed in the latest version. By the way, is there an advantage using `"` in this case, or is this more of a convention?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39993/32928

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #39993 was updated. @cmsbuild, @emeschi, @smorovic can you please check and sign again.

missirol commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c8184/28807/summary.html
**COMMIT**: d7db903445950bc684f769a3a710c75c165f10e1
**CMSSW**: CMSSW_12_6_X_2022-11-06-0000/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39993/28807/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416402
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416380
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-06-0000+0c8184/53920/triggerResults): no differences found

smorovic commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
